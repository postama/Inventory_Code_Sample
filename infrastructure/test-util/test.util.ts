import uuid from 'uuid/v4';

export function createBaseTestCommand(type) {
    return {
        commandId: uuid(),
        customerId: `TESTER`,
        user: 'abc',
        siteId: `TESTER`,
        type
    };
}