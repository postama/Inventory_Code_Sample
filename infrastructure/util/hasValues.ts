export function hasValues(object: object): boolean {
    return object && Object.keys(object).length > 0;
}

export function isEmpty(object: object): boolean {
    return !hasValues(object);
}