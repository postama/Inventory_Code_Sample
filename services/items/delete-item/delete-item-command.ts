import { type, TypeOf, literal, intersection } from 'io-ts';
import { validateType } from '../../../infrastructure/validator/validator';
import { BaseItemCommand } from '../shared/base-command-types';

export const DeleteItemCommandValidator = intersection([
    BaseItemCommand,
    type({ type: literal('DeleteItem')})
]);

export type DeleteItemCommand = TypeOf<typeof DeleteItemCommandValidator>;

export function createDeleteItemCommand(request: any, itemId: string): DeleteItemCommand {
    return validateType({ ...request, itemId, type: 'DeleteItem' }, DeleteItemCommandValidator);
}