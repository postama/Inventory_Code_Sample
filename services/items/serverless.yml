service: inventory-items

provider:
  name: aws
  runtime: nodejs8.10
  profile: kanso
  stage: local
  region: us-west-2
  environment:
    NODE_CONFIG_ENV: ${self:custom.stage}
  iamRoleStatements:
    - Effect: Allow
      Action:
        - dynamodb:Query
        - dynamodb:Scan
        - dynamodb:GetItem
        - dynamodb:PutItem
      Resource:
        - 'Fn::ImportValue': ${self:custom.stage}-InventoryTableArn
    - Effect: Allow
      Action:
        - dynamodb:Query
        - dynamodb:Scan
      Resource: {"Fn::Join": ["", [{"Fn::ImportValue":"${self:custom.stage}-InventoryTableArn"}, "/index/*"]]}

package:
  individually: ${self:custom.packageIndividually.${self:custom.stage}, self:custom.packageIndividually.other}

plugins:
  - serverless-add-api-key
  - serverless-webpack
  - serverless-offline
  - serverless-domain-manager

custom:
  apiKeys:
    - name: inventory-key
      value: ${ssm:/doorways/${self:custom.stage}/INVENTORY_API_KEY}
  customDomain:
    domainName: ${self:custom.domainPrefix.${self:custom.stage}, self:custom.domainPrefix.other}.doorways-services.net
    certificateName: '*.doorways-services.net'
    basePath: item
  domainPrefix:
    prod: inventory
    other: ${self:custom.stage}-inventory
  stage: ${opt:stage, self:provider.stage}
  region: ${opt:region, self:provider.region}
  packageIndividually:
    local: false
    other: true

functions:
  IssueItem:
    handler: issue-item/issue-item.entry
    events:
      - http:
          path: v1/{itemId}/command/issue
          method: POST
          private: true
  CreateItem:
    handler: create-item/create-item.entry
    events:
      - http:
          path: v1
          method: POST
          private: true
  EditItem: 
    handler: edit-item/edit-item.entry
    events:
      - http:
          path: v1/{itemId}
          method: PUT
          private: true
  DeleteItem:
    handler: delete-item/delete-item.entry
    events:
      - http:
          path: v1/{itemId}
          method: DELETE
          private: true
  AdjustItem:
    handler: adjust-qty/adjust-qty.entry
    events:
      - http:
          path: v1/{itemId}/command/adjust
          method: POST
          private: true
  ReceiveItem:
    handler: receive-item/receive-item.entry
    events:
      - http:
          path: v1/{itemId}/command/receive
          method: POST
          private: true
  DisposeItem:
    handler: dispose-item/dispose-item.entry
    events:
      - http:
          path: v1/{itemId}/command/dispose
          method: POST
          private: true
  TransferItem:
    handler: transfer-item/transfer-item.entry
    events:
      - http:
          path: v1/{itemId}/command/transfer
          method: POST
          private: true
