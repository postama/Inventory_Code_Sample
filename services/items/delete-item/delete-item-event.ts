import { type, literal, TypeOf, intersection } from 'io-ts';
import { DeleteItemCommand } from './delete-item-command';
import { validateType } from '../../../infrastructure/validator/validator';
import { createBaseEvent, BaseEventValidator } from '../../../repo/events/event-type';

export const DELETE_ITEM = 'ItemDeleted';

export const ItemDeletedValidatorV1 = intersection([
    BaseEventValidator,
    type({
        type: literal(DELETE_ITEM),
        version: literal(1)
    })
]);

export type ItemDeletedV1 = TypeOf<typeof ItemDeletedValidatorV1>;

export function createDeleteItemEvent(command: DeleteItemCommand): ItemDeletedV1 {
    const event = createBaseEvent(command, command.itemId, DELETE_ITEM, 1);
    return validateType(event, ItemDeletedValidatorV1);
}