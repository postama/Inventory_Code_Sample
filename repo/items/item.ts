import * as dynamo from '../../infrastructure/persistence/dynamo';
import { ItemEvent } from './item-event-types';
import { Item } from './item.type';
import { Reducer } from '../../infrastructure/reducer/reducer';
import { getEntityQuery } from '../events/entity';
import { CREATE_ITEM } from '../../services/items/create-item/create-item-event';
import { createItemReducer } from '../../services/items/create-item/create-item-reducer';
import { ADJUST_ITEM } from '../../services/items/adjust-qty/adjust-qty-event';
import { adjustQtyReducer } from '../../services/items/adjust-qty/adjust-qty-reducer';
import { ITEM_ISSUED } from '../../services/items/issue-item/issue-item-event';
import { issueItemReducer } from '../../services/items/issue-item/issue-item-reducer';
import { DISPOSE_ITEM } from '../../services/items/dispose-item/dispose-item-event';
import { disposeItemReducer } from '../../services/items/dispose-item/dispose-item-reducer';
import { RECEIVE_ITEM } from '../../services/items/receive-item/receive-item-event';
import { receiveItemReducer } from '../../services/items/receive-item/receive-item-reducer';
import { DELETE_ITEM } from '../../services/items/delete-item/delete-item-event';
import { deleteItemReducer } from '../../services/items/delete-item/delete-item-reducer';
import { EDIT_ITEM } from '../../services/items/edit-item/edit-item-event';
import { editItemReducer } from '../../services/items/edit-item/edit-item-reducer';
import { TRANSFER_ITEM } from '../../services/items/transfer-item/transfer-item-event';
import { transferItemReducer } from '../../services/items/transfer-item/transfer-item-reducer';

export const ItemReducer = new Reducer<Item, ItemEvent>();

// Code smell - breaking Open/Closed Principal these should be registered by their reducers, not in a central file.
// Not sure how to fix this - they need to be registered for every code split lambda, but how will each code split lambda bring in the code from the other events?
ItemReducer.register(CREATE_ITEM, createItemReducer);
ItemReducer.register(ADJUST_ITEM, adjustQtyReducer);
ItemReducer.register(ITEM_ISSUED, issueItemReducer);
ItemReducer.register(DISPOSE_ITEM, disposeItemReducer);
ItemReducer.register(RECEIVE_ITEM, receiveItemReducer);
ItemReducer.register(DELETE_ITEM, deleteItemReducer);
ItemReducer.register(EDIT_ITEM, editItemReducer);
ItemReducer.register(TRANSFER_ITEM, transferItemReducer);

export async function getById(customerId: string, itemId: string): Promise<Item> {
    return await dynamo.pagedReducer<Item, ItemEvent>(ItemReducer, getEntityQuery(customerId, itemId));
}

export async function save(event: ItemEvent) {
    return await dynamo.put(event);
}
