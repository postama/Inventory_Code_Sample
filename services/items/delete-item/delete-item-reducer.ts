import { Item } from '../../../repo/items/item.type';
import { ItemDeletedV1 } from './delete-item-event';

export function deleteItemReducer(item: Item, event: ItemDeletedV1) {
    return {
        ...item,
        deleted: event.created
    };
}