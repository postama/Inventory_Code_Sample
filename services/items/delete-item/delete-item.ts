import 'source-map-support/register';
import { APIGatewayEvent } from 'aws-lambda';
import { requestHandler } from '../../../../Kanso.Doorways.InventoryShared/entry-wrapper/entry-wrapper';
import { DeleteItemCommand, createDeleteItemCommand } from './delete-item-command';
import { checkUniqueCommand } from '../../../repo/events/event';
import { ItemDeletedV1, createDeleteItemEvent } from './delete-item-event';
import { deleteItemReducer } from './delete-item-reducer';
import { getById, save } from '../../../repo/items/item';

export async function entry(request: APIGatewayEvent) {
    return requestHandler(deleteItem, request, request.pathParameters.itemId);
}

export async function deleteItem(request: any, itemId: string): Promise<void> {
    const command: DeleteItemCommand = createDeleteItemCommand(request, itemId);
    await checkUniqueCommand<DeleteItemCommand>(command);
    const event: ItemDeletedV1 = createDeleteItemEvent(command);
    const item = await getById(event.customerId, event.entityId);
    event.state = deleteItemReducer(item, event);
    await save(event);
    return;
}