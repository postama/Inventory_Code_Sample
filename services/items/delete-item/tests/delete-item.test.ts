import uuid from 'uuid/v4';
import { createTestDeleteCommand, createTestDeleteEvent } from './test.util';
import { createDeleteItemCommand } from '../delete-item-command';
import { createDeleteItemEvent } from '../delete-item-event';
import { createTestCreateItemEvent } from '../../create-item/tests/test.util';
import { getById, save } from '../../../../repo/items/item';
import { entry, deleteItem } from '../delete-item';
import { deleteItemReducer } from '../delete-item-reducer';
import { createTestItem } from '../../../../repo/items/test/test-util';

import * as entryLib from '../../../../../Kanso.Doorways.InventoryShared/entry-wrapper/entry-wrapper';
jest.mock('../../../../../Kanso.Doorways.InventoryShared/entry-wrapper/entry-wrapper');
const entryMock = entryLib as jest.Mocked<typeof entryLib>;

describe('delete inventory item', () => {
    describe('create command', () => {
        test('should fail without commandId with a human readable error message', () => {
            const command = createTestDeleteCommand({ customerId: undefined });
            expect.assertions(2);
            try {
                createDeleteItemCommand(command, uuid());
            } catch (e) {
                expect(e.name).toBe('ValidationError');
                expect(e.message).toBe('Expecting string at customerId but instead got: undefined.');
            }
        });

        test('should create a command if provided proper structure', () => {
            const command = createTestDeleteCommand();
            const validCommand = createDeleteItemCommand(command, command.itemId);
            expect(validCommand).toEqual(command);
        });
    });

    describe('create event', () => {
        test('should fail if given a bad command', () => {
            expect.assertions(2);
            try {
                const command = createTestDeleteCommand({ commandId: undefined });
                createDeleteItemEvent(command);
            } catch (e) {
                expect(e.name).toBe('ValidationError');
                expect(e.message).toBe('Expecting string at commandId but instead got: undefined.');
            }
        });

        test('should create event if given a valid command', () => {
            const command = createTestDeleteCommand();
            const event = createDeleteItemEvent(command);
            expect(event.type).toBe('ItemDeleted');
        });
    });

    describe('delete event - integration', () => {
        test('should successfully remove the item', async () => {
            const event = createTestCreateItemEvent();
            await save(event);
            const deleteCommand = createTestDeleteCommand({ customerId: event.customerId });
            await deleteItem(deleteCommand, event.entityId);
            const item = await getById(event.customerId, event.entityId);
            expect(item).toMatchSnapshot({
                deleted: expect.any(String),
                created: expect.any(String)
            });
        });
    });

    describe('entry', () => {
        test('should call the entry handler', async () => {
            const command = createTestDeleteCommand();
            await entry({ body: JSON.stringify(command), pathParameters: { itemId: uuid() } } as any);
            expect(entryMock.requestHandler).toBeCalled();
        });
    });

    describe('reducer', () => {
        test('should add deleted to the item', async () => {
            const event = createTestDeleteEvent();
            const item = createTestItem();
            const deletedItem = deleteItemReducer(item, event);
            expect(deletedItem.deleted).toBe(event.created);
        });
    });
});