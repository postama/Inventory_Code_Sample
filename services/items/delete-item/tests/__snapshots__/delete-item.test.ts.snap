// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete inventory item delete event - integration should successfully remove the item 1`] = `
Object {
  "brand": "abc",
  "costPerUnit": 123,
  "created": Any<String>,
  "deleted": Any<String>,
  "group": "abc",
  "identifier": "abc",
  "isPreferred": true,
  "locations": Object {},
  "name": "abc",
  "quantity": 0,
  "supplier": "abc",
  "unitOfMeasurement": "abc",
}
`;
